### PR TITLE
Stafik/cloud archival/gc2

### PR DESCRIPTION
**Pull request by staffik**

### CODE REVIEW

File: chain/client/src/archive/cloud_archival_actor.rs [hunk 1]
```diff
@@ -169,32 +203,152 @@ impl CloudArchivalActor {
...
+/// Ensures `cloud_head` is not older than GC stop; returns `CloudHeadTooOld` otherwise.
+fn ensure_cloud_head_available_for_archiving(
+    hot_store: &Store,
+    runtime_adapter: &dyn RuntimeAdapter,
+    cloud_head: BlockHeight,
+) -> Result<(), CloudArchivalInitializationError> {
+    let block_hash = hot_store.chain_store().get_block_hash_by_height(cloud_head)?;
+    let gc_stop_height = runtime_adapter.get_gc_stop_height(&block_hash);
+    let gc_tail = hot_store.chain_store().tail()?;
+    if gc_tail > gc_stop_height {
```
Comment (Trisfald) [inline chain/client/src/archive/cloud_archival_actor.rs:305]: How do we prevent gc from moving while archival is uploading data?

I mean this corner case: can it happen that after this check and before data is uploaded, the GC has deleted some necessary blocks?

File: test-loop-tests/src/tests/cloud_archival.rs [hunk 2]
```diff
@@ -49,44 +73,58 @@ fn test_cloud_archival_base(params: TestCloudArchivalParameters) {
...
+        get_cloud_writer(&env, &archival_id).get_handle().resume();
+        stop_and_restart_node(&mut env, node_identifier.as_str());
+        gc_and_heads_sanity_checks(&env, &archival_id, params.enable_split_store, None);
+    }
+
+    let archival_node = TestLoopNode::for_account(&env.node_datas, &archival_id);
+    let target_height = params.num_epochs_to_wait * EPOCH_LENGTH;
+    archival_node.run_until_head_height(&mut env.test_loop, target_height);
+
+    println!("Final sanity checks");
```
Comment (Trisfald) [inline test-loop-tests/src/tests/cloud_archival.rs:92]: nit: move this println into `gc_and_heads_sanity_checks`?

File: test-loop-tests/src/utils/cloud_archival.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+use near_primitives::types::{AccountId, BlockHeight, BlockHeightDelta};
+use near_store::adapter::StoreAdapter;
+use near_store::db::CLOUD_HEAD_KEY;
+use near_store::{COLD_HEAD_KEY, DBCol};
+
+use crate::setup::env::TestLoopEnv;
+use crate::utils::node::TestLoopNode;
+
+/// Stops a node and restarts it with a new identifier `<old>-restart`.
+pub(crate) fn stop_and_restart_node(env: &mut TestLoopEnv, node_identifier: &str) {
```
Comment (Trisfald) [inline test-loop-tests/src/utils/cloud_archival.rs:12]: Is `pub(crate)` necessary if the module is `pub(crate) mod`?

File: chain/client/src/archive/cloud_archival_actor.rs [hunk 4]
```diff
@@ -46,26 +77,23 @@ pub struct CloudArchivalActor {
...
         return Ok(None);
     };
-    let hot_store = storage.get_hot_store();
-
-    // TODO(cloud_archival) Retrieve the `cloud_head` from the external storage
-    let cloud_head = hot_final_head_height(&hot_store, genesis_height)?;
+    let cloud_head =
+        initialize_cloud_head(&hot_store, genesis_height, &config.cloud_storage, runtime_adapter)?;
 
     tracing::info!(target: "cloud_archival", "Creating the cloud archival actor");
```
Comment (Trisfald) [inline chain/client/src/archive/cloud_archival_actor.rs:94]: consider add the `cloud_head` as a field (if it's not in some other log)

**Review by Trisfald — APPROVED** [review]
👍

File: test-loop-tests/src/tests/cloud_archival.rs (hunk 2 above)
Comment (staffik) [reply test-loop-tests/src/tests/cloud_archival.rs:92]: We call `gc_and_heads_sanity_checks` twice and the main purpose of this `println` is to differentiate between these two calls.

File: chain/client/src/archive/cloud_archival_actor.rs (hunk 1 above)
Comment (staffik) [reply chain/client/src/archive/cloud_archival_actor.rs:305]: It is prevented in `get_gc_stop_height_impl`. It is easier to review https://github.com/near/nearcore/pull/14350.
This PR is draft only.
